### DISCUSSION TITLE
Finer grained `loadPyodide`

### DISCUSSION THREAD

**Discussion by CNSeniorious000** [pyodide/pyodide#4706] [General]
Initiating Pyodide is unavoidable time-consuming. In my use cases, it usually consists of 3 steps:

1. loading the Pyodide files (about 5MB of wasm and stdlib zip files)
2. executing wasm (blocking about 1s)
3. installing micropip and other packages (relatively quick).

With browser caching, subsequent starts only require waiting for the wasm initialization block.

**However, the first step is particularly frustrating for me. Sometimes it even takes longer than 2 minutes under poor network conditions, and occasionally it fails to fetch the files.**

(This issue often arises in China where there is a firewall that blocks unregulated internet access, including public CDNs like jsDelivr.)

Therefore, I want to fetch these static assets myself, and provide the Blobs or as any format else to `loadPyodide`, instead of providing an `indexURL`, so that I can:

1. implement some retrying and racing logic to ensure the fetching process is predictable
2. show a finer-grained progress bar (currently I can only show a bar with 3 qualitative steps)
3. control the extract time to execute the wasm (which stops all the animations and blocks everything else)

Or, can `loadPyodide` provide some parameters like `loadPyodideWasm` and `loadStdlibZip`, or some lifecycle callbacks like `beforeExecutingWasm` and `onFetchTimeout`? I am not sure which one is the best practise.

**Comment by hoodmane** [comment]
Yeah we would be interested in doing something like this @ryanking13 has investigated the lifecycle callbacks approach e.g.
https://github.com/pyodide/pyodide/pull/3664

It's not that likely I'll have time to work on it any time soon but if you want to contribute something I'd be willing to help out with design/review.

**Reply by CNSeniorious000** [reply]
Yeah I am interested in contributing. Looks like the design of the plugin system is already done and what needs to do next is just list out what hooks should there be and inject some code into each function?

**Reply by ryanking13** [reply]
> design of the plugin system is already done and what needs to do next is just list out what hooks should there be and inject some code into each function?

Not really. Design considerations include clearly defining the initialization phase of Pyodide (currently it's a bit of a mess) and deciding what parts of it are exposed to the user. However, if you're interested in contributing, we're happy to help.

Of course, you don't have to do all of this. The draft PR I posted and the part you're interested in both involve preloading some files, so it would be good to define that first.
